ValveResourceFormat/IO/MapExtract.cs

```diff
@@ -359,10 +355,79 @@ public string ToValveMap()
             }
         }
 
+        //convert phys to hammer meshes
+        var phys = LoadWorldPhysics();
+        if (phys != null)
+        {
+            foreach (var hammermesh in PhysToHammerMeshes(phys))
+            {
+                MapDocument.World.Children.Add(new CMapMesh() { MeshData = hammermesh });
+            }
+        }
+
         using var stream = new MemoryStream();
+
+#if DEBUG
         datamodel.Save(stream, "keyvalues2", 4);
+#else
+        datamodel.Save(stream, "binary", 9);
+#endif
+
+        return stream.ToArray();
+    }
+
+    internal IEnumerable<CDmePolygonMesh> RenderMeshToHammerMesh(Model model, Resource resource, Vector3 offset = new Vector3())
+    {
+        var modelExtract = new ModelExtract(resource, FileLoader);
+        modelExtract.GrabMaterialInputSignatures(resource);
 
-        return Encoding.UTF8.GetString(stream.ToArray());
+        var dmxOptions = new ModelExtract.DatamodelRenderMeshExtractOptions
+        {
+            MaterialInputSignatures = modelExtract.MaterialInputSignatures,
+            SplitDrawCallsIntoSeparateSubmeshes = true,
+        };
+
+        foreach (var embedded in model.GetEmbeddedMeshes())
+        {
+            using var dmxMesh = ModelExtract.ConvertMeshToDatamodelMesh(embedded.Mesh, Path.GetFileNameWithoutExtension(resource.FileName), dmxOptions);
```

Comment:
A better approach IMO would be ModelExtract to give us the dmx objects. That allows it to handle external references (that are missed here) and input signatures.